src/checkpoint/dispatch/dispatch_virtual.h

```diff
@@ -48,4 +48,115 @@
 #include "checkpoint/dispatch/dispatch.h"
 #include "checkpoint/dispatch/vrt/virtual_serialize.h"
 
+/**
+ * ---------------------------------------------------------------------------
+ *
+ * Common header file for serializing virtual hierarchies of classes. Currently,
+ * only linear, single inheritance patterns are supported in the implementation.
+ *
+ * ---------------------------------------------------------------------------
+ * ----------------------------- API usage: ----------------------------------
+ * ---------------------------------------------------------------------------
+ *
+ * === Option 1 ===
+ *
+ *   - Make your virtual class hierarchy you want to serialize all inherit from
+ *     \c SerializableBase<T> and \c SerializableDervived<T,BaseT> in the whole
+ *     hierarchy.
+ *
+ * === Option 2 ===
+ *
+ *   - Insert checkpoint macros in your virtual class hierarchy for derived and
+ *     base classes with the corresponding macros:
+ *     - \c checkpoint_virtual_serialize_base(T)
+ *     - \c checkpoint_virtual_serialize_dervied(T,BaseT)
+ *
+ * Invoking the virtual serializer:
+ *
+ *   - If you have a \c std::unique_ptr<T>, where T is virtually serializable (by
+ *     using the macros or inheriting from \c SerializableBase<T> and
+ *     \c SerializableDervived<T,BaseT> ), they will automatically be virtually
+ *     serialized.
+ *
+ *   - If you have a raw pointer, \c Teuchos::RCP, or \c std::shared_ptr<T>,
+ *     you must invoke: \c checkpoint::allocateConstructForPointer<SerializerT,T>
+ *
+ *     Example with raw pointer:
+ *
+ *       template <typename T>
+ *       struct MyObjectWithRawPointer {
+ *         T* raw_ptr;
+ *
+ *         template <typename SerializerT>
+ *         void serialize(SerializerT& s) {
+ *           bool is_null = raw_ptr == nullptr;
+ *           s | is_null; // serialize whether we have a null pointer not
+ *           if (!is_null) {
+ *             // During size/pack, save the actual derived type of raw_ptr;
+ *             // During unpack, allocate/construct raw_ptr with correct virtual type
+ *             checkpoint::allocateConstructForPointer(s, raw_ptr);
+ *             s | *raw_ptr;
+ *           }
+ *         }
+ *       };
```

Comment:
Could we extract this motif into a function that any of these pointer holders can easily call?

```c++
template <typename SerializerT, typename T>
void serializeOwningPointer(SerializerT& s, T*& raw_ptr) {
  bool is_null = raw_ptr == nullptr;
  s | is_null; // serialize whether we have a null pointer not
  if (!is_null) {
    // During size/pack, save the actual derived type of raw_ptr;
    // During unpack, allocate/construct raw_ptr with correct virtual type
    checkpoint::allocateConstructForPointer(s, raw_ptr);
    s | *raw_ptr;
  }
}
```

Reply:
Hmm, yes I think we could: In fact, we can provide some overloads on top of it for things like shared_ptr, or maybe even this:

```C++
template <typename SerializerT, template <typename T> class SmartPtrLike>
void serializeOwningPointer(SerializerT& s, SmartPtrLike<T>& ptr) {
  // call down
}
```

Reply:
I'd be really hesitant to write something that attempts to be generic across smart pointer types. It's a pretty broad field of possibilities, and we don't know what's going to break if we get it wrong.

I'd lean toward just making it very easy to implement explicit functions for specific smart pointer types, and providing this overload as suggested only for `shared_ptr`, and maybe with library availability checking for `Teuchos::RCP`

Reply:
If you're not already into it, I should say that this doesn't need to be part of this PR. It's an unrelated convenience.